### PR TITLE
Fix type error when queryOptions enable is set to false

### PR DESCRIPTION
**Pull request by yanchesky**
Fixes https://github.com/marmelab/react-admin/issues/7986

### CODE REVIEW

**Comment by djhi** [comment]
Thanks for the error report and the PR! Would you mind trying to add a unit test for this case?

File: packages/ra-core/src/controller/list/useListController.ts [hunk 1]
```diff
@@ -109,7 +109,9 @@ export const useListController = <RecordType extends RaRecord = any>(
     useEffect(() => {
         if (
             query.page <= 0 ||
-            (!isFetching && query.page > 1 && data.length === 0)
+            (!isFetching &&
+                query.page > 1 &&
+                (data === undefined || data.length === 0))
```
Comment (djhi) [inline packages/ra-core/src/controller/list/useListController.ts:114]:
```suggestion
                (data == null || data.length === 0))
```

Might be even safer

File: packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx [hunk 2]
```diff
@@ -212,7 +212,7 @@ export const Datagrid: FC<DatagridProps> = React.forwardRef((props, ref) => {
      * displaying the table header with zero data rows,
      * the datagrid displays nothing or a custom empty component.
      */
-    if (data.length === 0 || total === 0) {
+    if (data === undefined || data.length === 0 || total === 0) {
```
Comment (djhi) [inline packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx:215]:
```suggestion
    if (data == null || data.length === 0 || total === 0) {
```
same

**Review by djhi — CHANGES_REQUESTED** [review]
Thanks!

**Review by djhi — APPROVED** [review]
Thanks!

**Comment by yanchesky** [comment]
Fixing failed unit test might be cumbersome 🤔. I have run it on the local machine and out of 10 tests on unchanged code:

- 4 times all tests passed
- 3 times "should notify if passed an error with a message that makes the authProvider throw" test failed
- 2 times "should take only last change in case of a burst of changes (case of inputs being currently edited)" test failed
- 1 time both 👆 failed

**Comment by yanchesky** [comment]
I couldn't find the exact reason why the test `should take only last change in case of a burst of changes (case of inputs being currently edited)` sometimes fails. I've run this test only and it never failed. I suspect that it may be related to `react-router-dom` because right before the test fails, the error is raised `An update to HistoryRouter inside a test was not wrapped in act`. However, this error occurs even if the test passes so I'm not sure about this. 
However, knowing that I've come up with 2 solutions that make this test suite work. (By work I mean 10 test runs without any fail)

1. Create `history` and pass it to the CoreAdminContext component.
2. pass `disableSyncWithLocation` prop to the ListController component.

If that is ok with you I will make a commit fixing it.

 Snapshot of the mock requests that cause the test to fail:
```
"calls": Array [
    Array [
      "posts.listParams",
      Object {
        "filter": Object {},
        "order": "ASC",
        "page": 1,
        "perPage": 10,
        "sort": "id",
      },
    ],
    Array [
      "posts.listParams",
      Object {
        "filter": Object {
          "q": "hello",
        },
        "order": "ASC",
        "page": 1,
        "perPage": 10,
        "sort": "id",
      },
    ],
  ],
  ```

**Comment by ITopGun** [comment]
Hi

**Comment by WiXSL** [comment]
Could you resolve the conflict?

File: packages/ra-core/src/controller/list/useListController.ts [hunk 3]
```diff
@@ -109,6 +109,7 @@ export const useListController = <RecordType extends RaRecord = any>(
     useEffect(() => {
         if (
             query.page <= 0 ||
+            data == null ||
```
[the remark this answers is not in the document]
Comment (yanchesky) [reply packages/ra-core/src/controller/list/useListController.ts:112]: Thanks for the info. I will check it soon.

[the remark this answers is not in the document]
Comment (yanchesky) [reply packages/ra-core/src/controller/list/useListController.ts]: Changed it as you suggested and seems it's working now. However on my machine, some tests are failing, but they are failing on the master branch as well. 
unit test that fail: "should apply validation to both itself and its inner inputs"
e2e tests that fail: "should validate ArrayInput", and sometimes "should have tabbable menu items"

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
Thanks!

**Pull request merged**